### PR TITLE
skills: live async SkillSource, SkillCatalog, leaner framing; v0.9.12

### PR DESCRIPTION
**Pull request by cymoo**
## What

Ground-up rework of the skill system (per review discussion; intentionally breaking, no deprecation shims — pre-1.0 and `LocalDirSkillSource` was the only implementation).

### Protocol goes async + live — the reload question dissolves

\`SkillSource\` is now two async methods, each returning *current* truth:

```python
@runtime_checkable
class SkillSource(Protocol):
    async def list_skills(self) -> list[SkillMetadata]: ...
    async def load_skill(self, name: str) -> Skill: ...
```

There is deliberately **no reload/rescan seam**. \`DirectorySkillSource\` (was \`LocalDirSkillSource\`) re-scans per listing (worker thread, ~ms) and self-heals on a load miss — an agent that just wrote a new skill can \`load_skill\` it immediately; the prompt index refreshes at the next run's setup. DB/API/MCP backends fit naturally (query per call). Scan warnings are deduplicated once-per-process so per-run scans don't spam logs.

### Leaner injection framing (token cost)

The treat-as-data preamble used to ride along in **every** \`load_skill\` result and persist in the transcript. It now lives in the two tool descriptions (sent once, prompt-cacheable); results carry only the BEGIN/END markers. \`read_skill_file\` output is now framed + marker-neutralized exactly like \`load_skill\` (it was raw before — same-origin content, same trust posture).

### Naming & structure

- \`SkillCategory\` → \`SkillCatalog\` (its own docstrings already said "catalog"), with public \`list_skills()\` / \`load_skill()\`
- \`Skill.content\` → \`Skill.body\`; source/catalog/tool all share the \`list_skills\`/\`load_skill\` vocabulary
- new \`SkillNotFoundError(name, available=…)\` centralizes unknown-name errors; a filtered catalog rebuilds hints so hidden skill names never leak
- \`skills.py\` (832 lines) → \`skills/\` package: \`model\` / \`source\` / \`catalog\` / \`plugin\` (same import path)

### Fixes folded in

- Frontmatter opening delimiter must be \`---\` on its own line: a frontmatter-less body starting with a Markdown thematic break (\`----\`) is no longer silently swallowed
- Index rendering collapses newlines and caps extra values — frontmatter can no longer inject structure into the system prompt
- Scan-time IO moved out of the constructor; missing/unreadable roots are logged

## Testing

- \`pytest\`: 2119 passed, 31 skipped (skills suite rewritten: 123 tests, incl. new coverage for live pickup, self-heal, frontmatter edge cases, framing, filter-hint leaks, index sanitization)
- \`ruff check\` / \`ruff format\` clean on touched files; \`mypy lovia\` (strict) clean
- Docs updated (en/zh skills.md, architecture.md); external callers (\`Skills(*dirs)\` in web CLI, examples) unchanged

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_016B4Q3LoaTprKEHqEnByuzV

### CODE REVIEW

File: lovia/plugins/skills/source.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+    Tolerates leading blank lines and trailing whitespace on the ``---``
+    delimiters. If there is no closing delimiter the file has no valid
+    frontmatter, so it is treated entirely as body — guessing with a naive
+    substring search would truncate any value that contains "---". Malformed
+    YAML (or YAML that is not a mapping) yields ``{}`` so the caller's
+    name/description validation produces the actual error.
+    """
+
+    trimmed = text.lstrip()
```
[the remark this answers is not in the document]
Comment (cymoo) [reply lovia/plugins/skills/source.py]: Fixed in d49aa11 — the parser now skips only a BOM and blank lines (`_LEADING_BLANK_LINES`); an indented `---` is no longer treated as an opening delimiter, mirroring the closing rule. Added tests for the indented opener, whitespace-only blank lines, and BOM.

File: lovia/plugins/skills/source.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+    def _warn_once(self, key: str, msg: str, *args: object) -> None:
+        """Log *msg* at WARNING the first time *key* is seen.
+
+        The live source re-scans on every run; without deduplication a
+        broken skill directory would spam the log once per run.
+        """
+        if key not in self._warned:
+            self._warned.add(key)
+            logger.warning(msg, *args)
```
[the remark this answers is not in the document]
Comment (cymoo) [reply lovia/plugins/skills/source.py]: Fixed in d49aa11 — `_warn_once` now guards the check-then-add with a `threading.Lock`, logging outside the lock.

File: lovia/plugins/skills/catalog.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+            Args:
+                name: The skill name.
+                relpath: Relative path inside the skill directory.
+            """
+            try:
+                skill = await self.load_skill(name)
+                content = await asyncio.to_thread(skill.read_file, relpath)
+            except SkillsError as exc:
+                return str(exc)
+            return f"[skill: {skill.name}  file: {relpath}]\n{_frame(content)}"
```
[the remark this answers is not in the document]
Comment (cymoo) [reply lovia/plugins/skills/catalog.py]: Fixed in d49aa11 — the header now renders `_one_line(relpath, 500)` (and `load_skill` does the same for the on-disk path, since directory names are arbitrary bytes too). Added a test with a newline-bearing filename proving the header stays a single line.

File: lovia/plugins/skills/model.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                skill_name=self.name,
+                path=relpath,
+                hint="Use a relative path inside the skill directory.",
+            ) from None
+        if not target.is_file():
+            raise SkillsError(
+                f"Skill file not found: {relpath}",
+                skill_name=self.name,
+                path=relpath,
+            )
```
[the remark this answers is not in the document]
Comment (cymoo) [reply lovia/plugins/skills/model.py:215]: Fixed in d49aa11 — now `{relpath!r}`, consistent with the sibling error messages; test asserts the error string is single-line.
